include/ensmallen_bits/nelder_mead/nelder_mead_impl.hpp

```diff
@@ -0,0 +1,241 @@
+  // Nelder-Mead simplex algorithm with adaptive parameters" by F. Gao and L.
+  // Han (2010).
+  if (adaptiveDefault)
+  {
+    alpha = 1;
+    beta = 1.0 + 2.0 / (double) dim;
+    gamma = 0.75 - 0.5 / (double) dim;
+    delta = 1.0 - 1.0 / (double) dim;
+  }
+
+  // Construct the initial simplex.
+  MatType simplex;
+  simplexer.Simplex(simplex, iterate, function);
+
+  arma::rowvec fx(dim + 1);
+  for (size_t i = 0; i < dim + 1; ++i)
+  {
+    fx(i) = function.Evaluate(simplex.col(i));
+    Callback::Evaluate(*this, function, simplex.col(i), fx(i), callbacks...);
+  }
+
+  // Get the indices that correspond to the ordering of the function values
+  // at the vertices. order(0) is the index in the simplex of the vertex
+  // with the lowest function value, and order(dim) is the index in the
+  // simplex of the vertex with the highest function value.
+  arma::uvec order = arma::sort_index(fx);
+
+  terminate |= Callback::BeginOptimization(*this, function, iterate,
+      callbacks...);
+  for (size_t i = 1; i != maxIterations && !terminate; ++i)
+  {
+    const MatType M = arma::mean(simplex.cols(order.subvec(0, dim - 1)), 1);
+    const MatType xref = (1.0 + alpha) * M - alpha * simplex.col(order(dim));
+    const ElemType fref = function.Evaluate(xref);
+    Callback::Evaluate(*this, function, xref, fref, callbacks...);
+
+    if (fref < fx(order(0)))
+    {
+      // Compute expansion.
+      const MatType xexp = (1.0 + alpha * beta) * M - alpha * beta *
+          simplex.col(order(dim));
+      const ElemType fexp = function.Evaluate(xexp);
+      Callback::Evaluate(*this, function, xexp, fexp, callbacks...);
+
+      // Shift values by one to the right and swap the last with the first.
+      const size_t fLowest = order(dim);
+      for (size_t d = dim; d > 0; --d)
+        order(d) = order(d - 1);
+      order(0) = fLowest;
+
+      if (fexp < fref)
+      {
+        simplex.col(order(0)) = xexp;
+        fx(order(0)) = fexp;
+      }
+      else
+      {
+        simplex.col(order(0)) = xref;
+        fx(order(0)) = fref;
+      }
+    }
+    else if (fref < fx(order(dim - 1)))
+    {
+      // Accept reflection point.
+      const size_t s = ShiftOrder(order, fx, fref);
+      simplex.col(order(s)) = xref;
+      fx(order(s)) = fref;
+    }
+    else
+    {
+      if (fref < fx(order(dim)))
+      {
+        // Outside contraction.
+        const MatType xoc = (1.0 + alpha * gamma) * M - alpha * gamma *
+            simplex.col(order(dim));
+        const ElemType foc = function.Evaluate(xoc);
+        Callback::Evaluate(*this, function, xoc, foc, callbacks...);
+
+        if (foc <= fref)
+        {
+          const size_t s = ShiftOrder(order, fx, fref);
+          simplex.col(order(s)) = xoc;
+          fx(order(s)) = foc;
+        }
+        else
+        {
+          shrink = true;
+        }
+      }
+      else
+      {
+        // Inside contraction.
+        const MatType xic = (1.0 - gamma) * M + gamma *
```

Comment:
Should this have `alpha` multiplied too?  I did the algebra in my head and I think it should, but I didn't double check and I have a headache right now... :smile: